src/useDocumentVisibility/useDocumentVisibility.ts

```diff
@@ -0,0 +1,34 @@
+import { useState } from 'react';
+import { isBrowser } from '../util/const';
+import { useMountEffect } from '../useMountEffect/useMountEffect';
+
+const isDocumentVisible = () => document.visibilityState === 'visible';
+
+/**
+ * Returns a boolean indicating whether `document.visibilityState` equals `visible`.
+ * @param initializeWithValue Whether to initialize hook with value on first
+ * render or postpone it to effects stage. SSR users should set it to `false`.
+ */
+export function useDocumentVisibility(initializeWithValue = true): boolean | undefined {
+  const shouldInitializeWithValue = isBrowser && initializeWithValue;
+  const [isVisible, setIsVisible] = useState(
+    shouldInitializeWithValue ? isDocumentVisible() : undefined
+  );
+
+  useMountEffect(() => {
+    // We don't need to define an extra `useCallback` for this function as it is only required
+    // inside this effect. We save memory this way.
+    // eslint-disable-next-line unicorn/consistent-function-scoping
+    const handleVisibilityChange = () => setIsVisible(isDocumentVisible());
+
+    document.addEventListener('visibilitychange', handleVisibilityChange, {
+      passive: true,
+    });
+
+    return () => {
+      document.removeEventListener('visibilitychange', handleVisibilityChange);
+    };
```

Comment:
As xobotyi suggested, it would be beneficial to use `useEventListener` here, because it already handles all of this logic. In general, we prefer reusing our own hooks as much as possible, because that way the resulting hooks become more maintainable and improvements in one hook lead to improvements in other hooks as well.

Reply:
All of our hooks do.

Also, I meant to only use `useWindowSize` as reference for implementing the setting of the initial value.

Reply:
Hey @ArttuOll with my last commit this hook now uses `useEventListener` to listen to events. However, I don't know how to make this SSR-friendly, as `document` is not available on the server. I thought about conditionally calling the hook, but this breaks the rules. I think going with my initial idea (manually registering events) would be the best option in this case, what do you think?

Reply:
Do not reference web-apis during SSR and everything will be fine🙃
i gave you solution here: https://github.com/react-hookz/web/pull/1049#discussion_r1051625771

In case we're not in browser (therefore were in ssr) - `document` not referenced, effects are not invoked during ssr therefore there is no need in guarding there.